packages/babel-core/src/transformation/internal-plugins/shadow-functions.js

```diff
@@ -74,11 +88,19 @@ function remap(path, key, create) {
   let cached = fnPath.getData(key);
   if (cached) return path.replaceWith(cached);
 
-  let init = create();
   let id   = path.scope.generateUidIdentifier(key);
 
   fnPath.setData(key, id);
-  fnPath.scope.push({ id, init });
+
+  if (key === "this" && fnPath.isMethod({kind: "constructor"})){
```

Comment:
We should add `"object-curly-spacing": ["error", "always"]` too for `({ id })` vs `({kind: "constructor"})`?